cmd/cmd.go

```diff
@@ -70,3 +80,503 @@ func validateUserConfig(cfg *viper.Viper) error {
+
+	// duck-type for downloadParams created from varying types
+	downloadableFrom
+}
+
+// newDownloadParamsFromExercise creates a new downloadParams given an exercise.
+func newDownloadParamsFromExercise(exercise ws.Exercise, usrCfg *viper.Viper) (*downloadParams, error) {
+	d := &downloadParams{
+		slug:             exercise.Slug,
+		track:            exercise.Track,
+		downloadableFrom: downloadableFromExercise{},
+	}
+	return d.buildDownloadParams(usrCfg)
+}
+
+// newDownloadParamsFromFlags creates a new downloadParams given flags.
+func newDownloadParamsFromFlags(flags *pflag.FlagSet, usrCfg *viper.Viper) (*downloadParams, error) {
+	d := &downloadParams{downloadableFrom: downloadableFromFlags{}}
+	var err error
+	d.uuid, err = flags.GetString("uuid")
+	if err != nil {
+		return nil, err
+	}
+	d.slug, err = flags.GetString("exercise")
+	if err != nil {
+		return nil, err
+	}
+	d.track, err = flags.GetString("track")
+	if err != nil {
+		return nil, err
+	}
+	d.team, err = flags.GetString("team")
+	if err != nil {
+		return nil, err
+	}
+	return d.buildDownloadParams(usrCfg)
+}
+
+// buildDownloadParams contains the common creation logic for creating downloadParams.
+func (d *downloadParams) buildDownloadParams(usrCfg *viper.Viper) (*downloadParams, error) {
+	d.token = usrCfg.GetString("token")
+	d.apibaseurl = usrCfg.GetString("apibaseurl")
+	d.workspace = usrCfg.GetString("workspace")
+	return d, d.validate()
+}
+
+// validate validates creation of downloadParams.
+func (d downloadParams) validate() error {
+	validator := downloadParamsValidator{params: &d}
+
+	if err := validator.needsSlugXorUUID(); err != nil {
+		return err
+	}
+	if err := validator.needsUserConfigValues(); err != nil {
+		return err
+	}
+	if err := validator.needsSlugWhenGivenTrackOrTeam(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ensureExerciseFilesWritable checks permission for writing exercise files.
+func (d downloadParams) ensureExerciseFilesWritable() error {
+	if !d.downloadableFrom.writeExerciseFilesPermitted() {
+		return errors.New("writing exercise files not permitted when downloading from this type")
+	}
+	return nil
+}
+
+// writeExerciseFilesPermitted is a template pattern default.
+func (d downloadParams) writeExerciseFilesPermitted() bool { return false }
+
+// errMissingSlugOrUUID is a template pattern default.
+func (d downloadParams) errMissingSlugOrUUID() error {
+	return errors.New("need a 'slug' or a 'uuid'")
+}
+
+// errGivenTrackOrTeamMissingSlug is a template pattern default.
+func (d downloadParams) errGivenTrackOrTeamMissingSlug() error {
+	return errors.New("track or team requires slug (not uuid)")
+}
+
+// downloadableFrom is the interface to use the template pattern when creating downloadParams from different types.
+// Clients can embed downloadParams to delegate to the default implementation.
+// This allows fine-grained specializations without having to define the entire interface.
+type downloadableFrom interface {
```

Comment:
I'm not sure how likely it is that there will be new data sources initiating a download. If it seems very unlikely, this is probably overkill. At present, it seems clear that we want slightly different behavior when downloading from Flags vs an Exercise.

If we imagine a future change allowing a download from ExerciseMetadata, it seems that the behavior and error messages might be the same as when downloading from an Exercise. If this sounds like a reasonable assumption then this interface might be simplified into a bool to signal creation from flags (as this is the only varying behavior). 

If we want to leave open the ability to add custom behavior when initiating a download from different types, then having an explicit duck type to easily hook into seems useful though.